pysph/base/octree.pyx

```diff
@@ -875,19 +944,373 @@ cdef class CompressedOctree(Octree):
+
+            new_node = self._new_node(xmin_current, length_padded,
+                    hmax=hmax_children[oct_id], level=1, parent=node)
+            node.children[oct_id] = new_node
+            new_node.start_index = c
+            new_node.num_particles = count[oct_id]
+            count[oct_id] = c
+            c = c + new_node.num_particles
+            if (new_node.num_particles < self.leaf_max_particles):
+                new_node.is_leaf = True
+                continue
+            next_level_nodes.push_back(new_node)
+
+        # Assign p_indices directly according to which child the index belongs to and
+        # where each child of root starts
+        with nogil, parallel():
+            for p in prange(n, schedule='static'):
+                tid = threadid()
+                oct_id = deref(child_indices)[p]
+                oct_cid = count[oct_id] + cumulative_map[tid][oct_id]
+                p_indices[oct_cid] = p
+                cumulative_map[tid][oct_id] += 1
+
+        del child_indices
+
+        if (next_level_nodes.empty()):
+            del next_level_nodes
+            return 2
+        else:
+            return self._c_build_tree_bfs(pa, p_indices, next_level_nodes, 1, num_threads)
+
+
+    @cython.cdivision(True)
+    @cython.boundscheck(False)
+    cdef int _c_build_tree_bfs(self, NNPSParticleArrayWrapper pa, u_int* p_indices,
+             vector[cOctreeNode *]* level_nodes, int level, int num_threads) nogil:
+
+        cdef double* src_x_ptr = pa.x.data
+        cdef double* src_y_ptr = pa.y.data
+        cdef double* src_z_ptr = pa.z.data
+        cdef double* src_h_ptr = pa.h.data
+
+        cdef double eps, length_padded, length, x_length, y_length, z_length, length_new
+        cdef cOctreeNode* node
+        cdef int i, j, k, n, num_nodes, l, start, num_p, oct_id, sid, jid, tid, r, d, children
+        cdef u_int p, q
+        cdef vector[double] hmax_children
+        cdef vector[vector[u_int]] new_indices
+        cdef double *xmin
+        cdef vector[dbl_ptr] xmax_new
+        cdef vector[dbl_ptr] xmin_new
+        cdef double *xmin_current
+        cdef double *xmax_current
+
+        cdef vector[vector[node_ptr]] old_nodes = vector[vector[node_ptr]](num_threads)
+        cdef vector[vector[node_ptr]] new_nodes = vector[vector[node_ptr]](num_threads)
+
+        for i from 0<=i<num_threads:
+            old_nodes[i] = vector[node_ptr]()
+            new_nodes[i] = vector[node_ptr]()
+
+        cdef vector[int] count = vector[int](num_threads+2)
+        d = 0
+        num_nodes = level_nodes.size()
+        for i from 0<=i<num_nodes:
+            count[d] = count[d] + 1
+            old_nodes[d].push_back(deref(level_nodes)[i])
+            d = (d + 1)%num_threads
+        count[num_threads] = num_nodes
+        count[num_threads + 1] = 2
+
+        del level_nodes
+
+        with nogil, parallel():
+            xmin_current = <double *> malloc(3 * sizeof(double))
```

Comment:
Do you need to malloc this? Can you not allocate this on the stack with a `double[3]`?  I ask because I don't see a matching free in this PR or in the sources.  For every malloc there should be a matching free. Similarly for every "new" a corresponding "del".  I think there may still lurk some issues like this.